src/System.Runtime.Analyzers/Core/SpecifyStringComparison.cs

```diff
@@ -24,15 +29,117 @@ public abstract class SpecifyStringComparisonAnalyzer : DiagnosticAnalyzer
+
+                // Without these symbols the rule cannot run
+                if (stringComparisonType == null || stringType == null || currentCulture == null ||
+                    currentCultureIgnoreCase == null || ordinalIgnoreCase == null || ordinal == null)
+                {
+                    return;
+                }
+
+                var stringCompareToNamedMethods = stringType.GetMembers("CompareTo").OfType<IMethodSymbol>();
+                var stringCompareToParameterString = stringCompareToNamedMethods.GetSingleOrDefaultMemberWithName("string.CompareTo(string)");
+                var stringCompareToParameterObject = stringCompareToNamedMethods.GetSingleOrDefaultMemberWithName("string.CompareTo(object)");
+
+                var boolString = csaContext.Compilation.Language == LanguageNames.CSharp ? "bool" : "Boolean";
+                var intString = csaContext.Compilation.Language == LanguageNames.CSharp ? "int" : "Integer";
+                var stringCompareNamedMethods = stringType.GetMembers("Compare").OfType<IMethodSymbol>();
+                var stringCompareParameterStringStringBool = stringCompareNamedMethods.GetSingleOrDefaultMemberWithName($"string.Compare(string, string, {boolString})");
+                var stringCompareParameterStringStringComparison = stringCompareNamedMethods.GetSingleOrDefaultMemberWithName("string.Compare(string, string, System.StringComparison)");
+                var stringCompareParameterStringIntStringIntIntBool = stringCompareNamedMethods.GetSingleOrDefaultMemberWithName($"string.Compare(string, {intString}, string, {intString}, {intString}, {boolString})");
+                var stringCompareParameterStringIntStringIntIntComparison = stringCompareNamedMethods.GetSingleOrDefaultMemberWithName($"string.Compare(string, {intString}, string, {intString}, {intString}, System.StringComparison)");
+
+                IDictionary<IMethodSymbol, IMethodSymbol> overloadMap = new Dictionary<IMethodSymbol, IMethodSymbol>();
+                overloadMap.AddKeyValueIfNotNull(stringCompareToParameterString, stringCompareParameterStringStringComparison);
+                overloadMap.AddKeyValueIfNotNull(stringCompareToParameterObject, stringCompareParameterStringStringComparison);
+                overloadMap.AddKeyValueIfNotNull(stringCompareParameterStringStringBool, stringCompareParameterStringStringComparison);
+                overloadMap.AddKeyValueIfNotNull(stringCompareParameterStringIntStringIntIntBool, stringCompareParameterStringIntStringIntIntComparison);
+
+                csaContext.RegisterOperationAction(oaContext =>
+                {
+                    var invocationExpression = (IInvocationExpression)oaContext.Operation;
+                    var targetMethod = invocationExpression.TargetMethod;
+
+                    if (targetMethod.IsGenericMethod ||
+                        targetMethod.ContainingType == null)
```

Comment:
Ah, I see that the logic below is not specific to String.CompareXXX methods, ignore my last comment.